### PR TITLE
[WIP][SPARK-43097] New pyspark ML logistic regression estimator implemented on top of distributor

### PR DESCRIPTION
**Pull request by WeichenXu123**

### CODE REVIEW

File: python/pyspark/mlv2/classification/logistic_regression.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        num_train_workers = self.getNumTrainWorkers()
+        batch_size = self.getBatchSize()
+
+        # Q: Shall we persist the shuffled dataset ?
+        # shuffling results are already cached
+        dataset = (
+            dataset
+            .select(self.getFeaturesCol(), self.getLabelCol())
+            .repartition(num_train_workers)
+            .persist()
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/classification/logistic_regression.py:147]: the dataset is used only twice?
1, to compute `num_rows` and `num_labels`;
2, write to local arrow files.

If so, I think we don't need to cache it.

File: python/pyspark/mlv2/classification/logistic_regression.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        for x, target in data_loader:
+            optimizer.zero_grad()
+            output = ddp_model(x)
+            loss_fn(output, target).backward()
+            optimizer.step()
+
+        # TODO: early stopping
+        #  When each epoch ends, computes loss on validation dataset and compare
+        #  current epoch validation loss with last epoch validation loss, if
+        #  less than provided `tol`, stop training.
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/classification/logistic_regression.py:126]: do we need to store the objective curve here?  or users can see the curve in tensorboard?

File: python/pyspark/mlv2/base.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+from pyspark.ml.param import P
+from pyspark.ml.common import inherit_doc
+from pyspark.ml.param.shared import (
+    HasInputCol,
+    HasOutputCol,
+    HasLabelCol,
+    HasFeaturesCol,
+    HasPredictionCol,
+)
+from pyspark.sql.dataframe import DataFrame
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/base.py:49]: will we have a `GenericDataFrame`? @HyukjinKwon 

I feel here should be something like:
```
if not is_remote():
    from pyspark.sql.dataframe import DataFrame
else:
    from pyspark.sql.connect.dataframe import DataFrame
```

Comment (HyukjinKwon) [reply python/pyspark/mlv2/base.py:49]: Yeah .. I am thinking that we might need a parent class for both too .. 
BTW, cannot call `is_remote()` when importing because `SPARK_REMOTE` is set when Spark session is created.

If this is just being used for type hinting, I think just using `from pyspark.sql.dataframe import DataFrame` would be fine for now because they have the same API anyway.

Comment (zhengruifeng) [reply python/pyspark/mlv2/base.py:49]: > I am thinking that we might need a parent class for both too
+1, FYI the new ML implementations will only use pure Dataframe/Column/Functions APIs, in order to support both regular spark and spark connect.

File: python/pyspark/mlv2/classification/logistic_regression.py (hunk 2 above)
Comment (WeichenXu123) [reply python/pyspark/mlv2/classification/logistic_regression.py:126]: we can add objective history in following PR.

File: python/pyspark/mlv2/base.py (hunk 3 above)
Comment (WeichenXu123) [reply python/pyspark/mlv2/base.py:49]: Not a big issue, the DataFrame class is only used for type hint in the code

File: python/pyspark/mlv2/classification/logistic_regression.py (hunk 1 above)
Comment (WeichenXu123) [reply python/pyspark/mlv2/classification/logistic_regression.py:147]: @zhengruifeng 

My thoughts:

In spark sql side, can we support a new kind of spark dataframe reader, that loads a spark dataframe from saved parquets/delta files, but the loaded dataframe is roughly evenly distributed with specific partition number, so that we can avoid the "repartition" (shuffle) step

File: python/pyspark/ml/torch/tests/test_data_loader.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+import sys
+import time
+import tempfile
+import threading
+import numpy as np
+from typing import Callable, Dict, Any
+import unittest
+from unittest.mock import patch
+
+have_torch = True
```
Comment (jaceklaskowski) [inline python/pyspark/ml/torch/tests/test_data_loader.py:33]: nit: Should this really be between `import`s?

File: python/pyspark/ml/torch/tests/test_data_loader.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+from pyspark import SparkConf, SparkContext
+from pyspark.ml.torch.distributor import TorchDistributor, get_gpus_owned, get_spark_partition_data_loader
+from pyspark.ml.torch.torch_run_process_wrapper import clean_and_terminate, check_parent_alive
+from pyspark.sql import SparkSession
+from pyspark.testing.utils import SPARK_HOME
+from pyspark.ml.linalg import Vectors
+
+
+@unittest.skipIf(not have_torch, "torch is required")
```
Comment (jaceklaskowski) [inline python/pyspark/ml/torch/tests/test_data_loader.py:47]: I think we need a (stacked) decorator to do torch check and skip if not available.

File: python/pyspark/ml/torch/tests/test_data_loader.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+class TorchDistributorDataLoaderUnitTests(unittest.TestCase):
+    def setUp(self) -> None:
+        self.spark = (
+            SparkSession.builder
+            .master("local[1]")
```
Comment (jaceklaskowski) [inline python/pyspark/ml/torch/tests/test_data_loader.py:52]: nit: indent right?

File: python/pyspark/ml/torch/tests/test_data_loader.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            .config("spark.default.parallelism", "1")
```
Comment (jaceklaskowski) [inline python/pyspark/ml/torch/tests/test_data_loader.py:53]: Is this required since `master("local[1]")`?

File: python/pyspark/mlv2/classification/base.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    HasProbabilityCol,
+    HasThresholds
+)
+
+
+class _ClassifierParams(HasRawPredictionCol, _PredictorParams):
+    """
+    Classifier Params for classification tasks.
+
+    .. versionadded:: 3.0.0
```
Comment (jaceklaskowski) [inline python/pyspark/mlv2/classification/base.py:31]: nit: 3.5.0 or even higher?

File: python/pyspark/mlv2/classification/logistic_regression.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
+import numpy as np
```
Comment (jaceklaskowski) [inline python/pyspark/mlv2/classification/logistic_regression.py:1]: nit: The ASF copyright header

**Comment by WeichenXu123** [comment]
New PR https://github.com/apache/spark/pull/41383
